akka-persistence-typed/src/test/scala/docs/akka/persistence/typed/AccountExampleWithOptionState.scala

```diff
@@ -0,0 +1,148 @@
+    final case class GetBalance()(override val replyTo: ActorRef[CurrentBalance])
+      extends AccountCommand[CurrentBalance]
+    final case class CloseAccount()(override val replyTo: ActorRef[OperationResult])
+      extends AccountCommand[OperationResult]
+
+    // Reply
+    sealed trait AccountCommandReply
+    sealed trait OperationResult extends AccountCommandReply
+    case object Confirmed extends OperationResult
+    final case class Rejected(reason: String) extends OperationResult
+    final case class CurrentBalance(balance: BigDecimal) extends AccountCommandReply
+
+    // Event
+    sealed trait AccountEvent
+    case object AccountCreated extends AccountEvent
+    case class Deposited(amount: BigDecimal) extends AccountEvent
+    case class Withdrawn(amount: BigDecimal) extends AccountEvent
+    case object AccountClosed extends AccountEvent
+
+    val Zero = BigDecimal(0)
+
+    // type alias to reduce boilerplate
+    type ReplyEffect = akka.persistence.typed.scaladsl.ReplyEffect[AccountEvent, Option[Account]]
+
+    // State
+    sealed trait Account {
+      def applyCommand(cmd: AccountCommand[_]): ReplyEffect
+      def applyEvent(event: AccountEvent): Account
+    }
+    case class OpenedAccount(balance: BigDecimal) extends Account {
+      require(balance >= Zero, "Account balance can't be negative")
+
+      override def applyCommand(cmd: AccountCommand[_]): ReplyEffect =
+        cmd match {
+          case c @ Deposit(amount) ⇒
+            Effect.persist(Deposited(amount))
+              .thenReply(c)(_ ⇒ Confirmed)
+
+          case c @ Withdraw(amount) ⇒
+            if (canWithdraw(amount)) {
+              Effect.persist(Withdrawn(amount))
+                .thenReply(c)(_ ⇒ Confirmed)
+
+            } else {
+              Effect.reply(c)(Rejected(s"Insufficient balance $balance to be able to withdraw $amount"))
+            }
+
+          case c: GetBalance ⇒
+            Effect.reply(c)(CurrentBalance(balance))
+
+          case c: CloseAccount ⇒
+            if (balance == Zero)
+              Effect.persist(AccountClosed)
+                .thenReply(c)(_ ⇒ Confirmed)
+            else
+              Effect.reply(c)(Rejected("Can't close account with non-zero balance"))
+
+          case _: CreateAccount ⇒ Effect.unhandled.thenNoReply(cmd)
+
+        }
+
+      override def applyEvent(event: AccountEvent): Account = event match {
+        case Deposited(amount) ⇒ copy(balance = balance + amount)
+        case Withdrawn(amount) ⇒ copy(balance = balance - amount)
+        case AccountClosed     ⇒ ClosedAccount
+        case _                 ⇒ throw new IllegalStateException(s"unexpected event [$event] in state [OpenedAccount]")
+      }
+
+      def canWithdraw(amount: BigDecimal): Boolean = {
+        balance - amount >= Zero
+      }
+
+    }
+    case object ClosedAccount extends Account {
+      override def applyCommand(cmd: AccountCommand[_]): ReplyEffect =
+        Effect.unhandled.thenNoReply(cmd)
```

Comment:
That will cause a timeout. Maybe better to replace all `unhandled` by a `Reject`. 

Which makes me think that `thenNoReply` should not receive a command. If `cmd` is a command requiring a reply, how can I ignore it? Letting the `Future` timeout can't be a good choice. 

That's very annoying. We can't safely decide if a reply is being expected.

Reply:
> That will cause a timeout. Maybe better to replace all unhandled by a Reject.

I can do that. Thought it would be fine with some simplification, "it's just an example".

For example not replying to anything else than CreateAccount in EmptyAccount is completely valid in my opinion, because it's invalid protocol usage (also in the face of multiple users).

> Which makes me think that thenNoReply should not receive a command.

true, not needed

>  If cmd is a command requiring a reply, how can I ignore it?

It can be other reasons why you can't reply immediately, e.g. some async interaction after or before persisting the event. You can pass around the cmd (or the replyTo: ActorRef) and reply later.

> That's very annoying. We can't safely decide if a reply is being expected.

It's not about that. It's about not forgetting making a decision.

Reply:
I removed the cmd parameter from noReply in upstream PR.
Added more rejection replies, see c6d9758